osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/periodicmeterreads/GetPeriodicMeterReadsGasCommandExecutor.java

```diff
@@ -292,10 +305,22 @@ private Date readCaptureTime(
               bufferedObjects.get(captureTimeIndex), "Clock from mbus interval extended register");
 
       final Date captureTime;
-      if (cosemDateTime.isDateTimeSpecified()) {
-        captureTime = cosemDateTime.asDateTime().toDate();
+      final DateTime bufferedDateTime = cosemDateTime == null ? null : cosemDateTime.asDateTime();
+      if (bufferedDateTime != null) {

```

Comment:
It feels a bit like a double null check here (due to both `cosemDateTime == null` and `bufferedDateTime != null`). Perhaps you could consider rewriting to something like (pseudocode):
```
...
if (cosemDateTime == null) {
  return this.determineCaptureTimeFromPreviousValue(...);
} else {
  return this.determineCaptureTimeFromBuffer(...);
}
...

private Date determineCaptureTimeFromPreviousValue(...)  throws  BufferedDateTimeValidationException {
// or something like calculateCaptureTimeFromPreviousValue?
...
}

private Date determineCaptureTimeFromBuffer(...) throws ProtocolAdapterException {
 // or something like getBufferedCaptureTime?
...
}

```
When refactored this way, the `no date was available...` comment could probably be removed as well, because it should then be clear from the `determineCaptureTimeFromPreviousValue` method name....

Reply:
It's indeed checked for null values too often. I've simplified the function.